### PR TITLE
Spring usage improvements

### PR DESCRIPTION
**Pull request by diega**

### CODE REVIEW

File: rskj-core/src/main/java/org/ethereum/facade/EthereumImpl.java [hunk 1]
```diff
@@ -54,53 +53,37 @@
...
- * @author Roman Mandeleil
- * @since 27.07.2014
- */
-// @Component
 public class EthereumImpl implements Ethereum {
 
     private static final Logger logger = LoggerFactory.getLogger("facade");
     private static final Logger gLogger = LoggerFactory.getLogger("general");
 
-    @Autowired
```
Comment (tinchou) [inline rskj-core/src/main/java/org/ethereum/facade/EthereumImpl.java]: This is awesome :clap:

File: rskj-core/src/main/java/org/ethereum/facade/EthereumImpl.java [hunk 2]
```diff
@@ -54,53 +53,37 @@
...
+        this.peerServer = peerServer;
+        this.programInvokeFactory = programInvokeFactory;
+        this.pendingState = pendingState;
+        this.config = config;
+        this.compositeEthereumListener = compositeEthereumListener;
+        this.receiptStore = receiptStore;
+        this.peerClient = peerClient;
     }
 
-    @PostConstruct
```
Comment (tinchou) [inline rskj-core/src/main/java/org/ethereum/facade/EthereumImpl.java]: Ah, just realized why you called `init` above. Looks good now!

File: rskj-core/src/main/java/org/ethereum/facade/EthereumImpl.java [hunk 3]
```diff
@@ -124,7 +107,6 @@ public void connect(InetAddress addr, int port, String remoteId) {
     @Override
     public void connect(final String ip, final int port, final String remoteId) {
         logger.info("Connecting to: {}:{}", ip, port);
-        final PeerClient peerClient = ctx.getBean(PeerClient.class);
```
Comment (tinchou) [inline rskj-core/src/main/java/org/ethereum/facade/EthereumImpl.java]: I'm worried about the lifespan here. Doesn't this instance a new `PeerClient` every time? But if we inject it via constructor, then we will always have the same `peerClient` instance.
I'm invoking @ajlopez here to check.

File: rskj-core/src/test/java/co/rsk/mine/MinerManagerTest.java [hunk 4]
```diff
@@ -358,4 +358,11 @@ public boolean isValid(Block block) {
             return true;
         }
     }
+
+    private static class RskImplForTest extends RskImpl {
```
Comment (tinchou) [inline rskj-core/src/test/java/co/rsk/mine/MinerManagerTest.java]: Nit: Maybe create a method instead?

**Comment by rskops** [comment]
Can one of the admins verify this patch?

File: rskj-core/src/main/java/co/rsk/Start.java [hunk 5]
```diff
@@ -69,14 +76,6 @@ public void startNode() throws Exception {
             CONFIG.setDiscoveryEnabled(Boolean.FALSE);
         }
 
-        Rsk rsk = RskFactory.createRsk(config);
-
-        Runtime.getRuntime().addShutdownHook(new Thread() {
-            @Override
-            public void run() {
-                ((ConfigurableApplicationContext) RskFactory.getContext()).close();
```
[the remark this answers is not in the document]
Comment (diega) [reply rskj-core/src/main/java/co/rsk/Start.java]: No, this has been moved to the `main` method as:
```java
ConfigurableApplicationContext ctx = new AnnotationConfigApplicationContext(DefaultConfig.class);
ctx.registerShutdownHook()
```

File: rskj-core/src/main/java/co/rsk/core/RskFactory.java [hunk 6]
```diff
@@ -22,111 +22,89 @@
...
+import org.springframework.context.annotation.Configuration;
 
-/**
- * Created by ajlopez on 3/3/2016.
- */
-@Component
+import java.util.stream.Collectors;
+
+@Configuration
+@ComponentScan("org.ethereum")
```
[the remark this answers is not in the document]
Comment (diega) [reply rskj-core/src/main/java/co/rsk/core/RskFactory.java]: Spring, by default, scans for beans definitions under the same package  as the declared configuration, with this [@ComponentScan](https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/context/annotation/ComponentScan.html) we tell him to scan another packages

File: rskj-core/src/main/java/co/rsk/core/RskFactory.java [hunk 7]
```diff
@@ -22,111 +22,89 @@
...
-
-    public static Rsk createRsk(Class userSpringConfig) {
-        return createRsk(SystemProperties.CONFIG, userSpringConfig);
-    }
-
-    public static Rsk createRsk(SystemProperties config, Class userSpringConfig) {
-        logger.info("Running {},  core version: {}-{}", config.genesisInfo(), config.projectVersion(), config.projectVersionModifier());
-        BuildInfo.printInfo();
-
-        if (config.databaseReset()){
```
[the remark this answers is not in the document]
Comment (diega) [reply rskj-core/src/main/java/co/rsk/core/RskFactory.java]: Outstanding catch!!! I'll fix it real soon

File: rskj-core/src/main/java/co/rsk/core/RskFactory.java [hunk 8]
```diff
@@ -22,111 +22,89 @@
...
-            logger.info("Database reset done");
-        }
-
-        return userSpringConfig == null ? createRsk(new Class[] {DefaultConfig.class}) :
-                createRsk(DefaultConfig.class, userSpringConfig);
-    }
-
-    public static Rsk createRsk(Class ... springConfigs) {
 
+    @Bean
```
[the remark this answers is not in the document]
Comment (diega) [reply rskj-core/src/main/java/co/rsk/core/RskFactory.java]: I can try with `@Bean(autowire = Autowire.BY_TYPE)`

File: rskj-core/src/main/java/co/rsk/core/RskFactory.java [hunk 9]
```diff
@@ -22,111 +22,89 @@
...
+                      PeerServer peerServer,
+                      ProgramInvokeFactory programInvokeFactory,
+                      PendingState pendingState,
+                      SystemProperties config,
+                      CompositeEthereumListener compositeEthereumListener,
+                      ReceiptStore receiptStore,
+                      PeerClient peerClient) {
+        RskImpl rsk = new RskImpl(worldManager, adminInfo, channelManager, peerServer, programInvokeFactory,
+                pendingState, config, compositeEthereumListener, receiptStore, peerClient);
+        rsk.init();
```
[the remark this answers is not in the document]
Comment (diega) [reply rskj-core/src/main/java/co/rsk/core/RskFactory.java]: It was called by Spring b/c it had the `@PostConstruct`, as now I'm manually creating the instance I preferred to do it explicitly. The other option would have been to set it in the annotation as `@Bean(initMethod = "init")`

File: rskj-core/src/main/java/co/rsk/core/RskFactory.java [hunk 10]
```diff
@@ -22,111 +22,89 @@
...
+        rsk.getBlockchain().setRsk(true);  //TODO: check if we can remove this field from org.ethereum.facade.Blockchain
         if (logger.isInfoEnabled()) {
-            StringBuilder versions = new StringBuilder();
-            for (EthVersion v : EthVersion.supported()) {
-                versions.append(v.getCode()).append(", ");
-            }
-            versions.delete(versions.length() - 2, versions.length());
-            logger.info("capability eth version: [{}]", versions);
+            String versions = EthVersion.supported().stream().map(EthVersion::name).collect(Collectors.joining(", "));
```
[the remark this answers is not in the document]
Comment (diega) [reply rskj-core/src/main/java/co/rsk/core/RskFactory.java]: For logging purposes it made no difference and using the _name_ gave me an String to print directly, if I used the _code_ I would had to convert the byte to string.

File: rskj-core/src/main/java/org/ethereum/facade/EthereumImpl.java (hunk 3 above)
Comment (diega) [reply rskj-core/src/main/java/org/ethereum/facade/EthereumImpl.java]: Another great catch, I didn't realize the `@Scope("prototype")` of the **PeerClient**. I'll take another look tomorrow.

File: rskj-core/src/main/java/org/ethereum/facade/EthereumImpl.java (hunk 2 above)
Comment (diega) [reply rskj-core/src/main/java/org/ethereum/facade/EthereumImpl.java]: Cool

File: rskj-core/src/test/java/co/rsk/mine/MinerManagerTest.java (hunk 4 above)
Comment (diega) [reply rskj-core/src/test/java/co/rsk/mine/MinerManagerTest.java]: What's _nit_?

Comment (tinchou) [reply rskj-core/src/test/java/co/rsk/mine/MinerManagerTest.java]: https://stackoverflow.com/questions/27810522/what-does-nit-mean-in-hacker-speak

File: rskj-core/src/main/java/co/rsk/Start.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (tinchou) [reply rskj-core/src/main/java/co/rsk/Start.java]: Ah, I see. We were doing it completely wrong.

**Comment by diega** [comment]
There are at least three more **prototype** beans that needs to be replaced by _factories_, those are `org.ethereum.net.p2p.P2pHandler`, `org.ethereum.net.rlpx.MessageCodec` and `org.ethereum.net.MessageQueue`

File: rskj-core/src/test/java/co/rsk/mine/MinerManagerTest.java (hunk 4 above)
Comment (diega) [reply rskj-core/src/test/java/co/rsk/mine/MinerManagerTest.java]: We need this as a class b/c the test creates anonymous classes extending this one and overrides some methods

**Pull request merged**